src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs

```diff
@@ -106,41 +137,75 @@ private async Task<T> GetSystemDocumentByIdAsync<T>(string id, string partitionK
-                var result = response.SingleOrDefault();
-
-                if (result == null)
-                {
-                    _logger.LogError($"{typeof(T)} with id {id} was not found in Cosmos DB.");
-                }
-
-                return result;
+                FeedResponse<Document> response = await cosmosDocumentQuery.ExecuteNextAsync<Document>(cancellationToken);
+                return response;
             }
         }
 
-        private async Task<T> UpsertSystemObjectAsync<T>(T systemObject, string partitionKey, CancellationToken cancellationToken)
+        private async Task<UpsertResponse<T>> UpsertSystemObjectAsync<T>(T systemObject, string partitionKey, string eTag,  CancellationToken cancellationToken)
             where T : class
         {
             EnsureArg.IsNotNull(systemObject, nameof(systemObject));
+            RequestOptions requestOptions = GetRequestOptions(partitionKey, eTag);
+
+            try
+            {
+                var response = await _documentClient.Value.UpsertDocumentAsync(
+                    _collectionUri,
+                    systemObject,
+                    requestOptions,
+                    true,
+                    cancellationToken);
+                _logger.LogInformation("Request charge: {RequestCharge}, latency: {RequestLatency}", response.RequestCharge, response.RequestLatency);
+
+                var outcomeType = response.StatusCode == HttpStatusCode.Created ? UpsertOutcomeType.Created : UpsertOutcomeType.Updated;
+                return new UpsertResponse<T>((T)(dynamic)response.Resource, outcomeType, response.Resource.ETag);
+            }
+            catch (DocumentClientException dce)
+            {
+                _logger.LogError(dce, "Unhandled Document Client Exception");
+                throw;
+            }
+        }
+
+        private static RequestOptions GetRequestOptions(string partitionKey, string eTag)
+        {
             var eTagAccessCondition = new AccessCondition();
 
+            if (!string.IsNullOrWhiteSpace(eTag))
+            {
+                eTagAccessCondition.Type = AccessConditionType.IfMatch;
+                eTagAccessCondition.Condition = eTag;
+            }
+
             var requestOptions = new RequestOptions
             {
                 PartitionKey = new PartitionKey(partitionKey),
                 AccessCondition = eTagAccessCondition,
             };
+            return requestOptions;
+        }
+
+        private async Task DeleteSystemDocumentsByIdAsync<T>(string id, string partitionKey, string eTag, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(id, nameof(id));
+            FeedResponse<Document> response = await GetSystemDocuments<T>(id, partitionKey, cancellationToken);
+            var document = response.SingleOrDefault();
+
+            if (document == null)
+            {
+                throw new IdentityProviderNotFoundException(id);
```

Comment:
This should be a more general exception if being thrown in the `DeleteSystemDocumentsByIdAsync` method